yarn-project/simulator/src/public/fixtures/public_tx_simulation_tester.ts

```diff
@@ -0,0 +1,170 @@
+    const simulator = new PublicTxSimulator(this.merkleTrees, worldStateDB, globals, /*doMerkleOperations=*/ true);
+
+    const setupExecutionRequests: PublicExecutionRequest[] = [];
+    for (let i = 0; i < setupCalls.length; i++) {
+      const address = setupCalls[i].address || this.getFirstContractInstance().address;
+      const req = executionRequestForCall(
+        sender,
+        address,
+        setupCalls[i].fnName,
+        setupCalls[i].args,
+        setupCalls[i].isStaticCall,
+        setupCalls[i].contractArtifact,
+      );
+      setupExecutionRequests.push(req);
+    }
+    const appExecutionRequests: PublicExecutionRequest[] = [];
+    for (let i = 0; i < appCalls.length; i++) {
+      const address = appCalls[i].address || this.getFirstContractInstance().address;
+      const req = executionRequestForCall(
+        sender,
+        address,
+        appCalls[i].fnName,
+        appCalls[i].args,
+        appCalls[i].isStaticCall,
+        appCalls[i].contractArtifact,
+      );
+      appExecutionRequests.push(req);
+    }
+
+    let teardownExecutionRequest: PublicExecutionRequest | undefined = undefined;
+    if (teardownCall) {
+      const address = teardownCall.address || this.getFirstContractInstance().address;
+      teardownExecutionRequest = executionRequestForCall(
+        sender,
+        address,
+        teardownCall.fnName,
+        teardownCall.args,
+        teardownCall.isStaticCall,
+        teardownCall.contractArtifact,
+      );
+    }
+
+    const tx: Tx = createTxForPublicCalls(
+      setupExecutionRequests,
+      appExecutionRequests,
+      teardownExecutionRequest,
+      feePayer,
+    );
+
+    const avmResult = await simulator.simulate(tx);
+
+    if (avmResult.revertCode.isOK()) {
+      await this.commitTxStateUpdates(avmResult.avmProvingRequest.inputs.output);
+    }
+
+    return avmResult;
+  }
+
+  public async simulateAppLogicOnlyTx(appCall: TestEnqueuedCall) {
+    return await this.simulateTx(
+      /*setupCalls=*/ [],
+      /*appCalls=*/ [appCall],
+      /*teardownCall=*/ undefined,
+      /*sender=*/ undefined,
+      /*feePayer=*/ undefined,
+    );
+  }
+
+  private async commitTxStateUpdates(avmCircuitInputs: AvmCircuitPublicInputs) {
+    await this.merkleTrees.appendLeaves(
+      MerkleTreeId.NOTE_HASH_TREE,
+      padArrayEnd(avmCircuitInputs.accumulatedData.noteHashes, Fr.ZERO, MAX_NOTE_HASHES_PER_TX),
+    );
+    try {
+      await this.merkleTrees.batchInsert(
+        MerkleTreeId.NULLIFIER_TREE,
+        padArrayEnd(avmCircuitInputs.accumulatedData.nullifiers, Fr.ZERO, MAX_NULLIFIERS_PER_TX).map(n => n.toBuffer()),
+        NULLIFIER_SUBTREE_HEIGHT,
+      );
+    } catch (error) {
+      this.logger.warn(`Detected duplicate nullifier.`);
+    }
+
+    await this.merkleTrees.batchInsert(
+      MerkleTreeId.PUBLIC_DATA_TREE,
+      avmCircuitInputs.accumulatedData.publicDataWrites.map(w => w.toBuffer()),
+      PUBLIC_DATA_TREE_HEIGHT,
+    );
+  }
+}
```

Comment:
Nice.

I wonder if the TXE has a similar logic somewhere where it does all this for a TX?